http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala

```diff
@@ -0,0 +1,342 @@
+    allRuntimes: List[RuntimeMetrics]
+  )(implicit ev: Ask[F, AppContext]): F[Map[RuntimeHealthMetric, Int]] = {
+    val allContainers = for {
+      // Only care about Running runtimes for health checks
+      r <- allRuntimes if r.status == RuntimeStatus.Running
+      c <- r.containers
+    } yield (r, c)
+
+    allContainers
+      .parTraverseN(parallelism) { case (runtime, container) =>
+        for {
+          ctx <- ev.ask
+          isUp <- container.isProxyAvailable(runtime.cloudContext, runtime.runtimeName)
+          // In addition to collecting aggregate metrics, log a warning for any runtime that is down.
+          _ <-
+            if (isUp) F.unit
+            else
+              logger.error(ctx.loggingCtx)(
+                s"Runtime is DOWN with " +
+                  s"name={${runtime.runtimeName.asString}}, " +
+                  s"type={${container.imageType.toString}}, " +
+                  s"cloudContext={${runtime.cloudContext.asStringWithProvider}}, " +
+                  s"workspace={${runtime.workspaceId}}"
+              )
+        } yield Map(
+          RuntimeHealthMetric(runtime.cloudContext.cloudProvider,
+                              container.imageType,
+                              getRuntimeUI(runtime.labels),
+                              getAzureCloudContext(runtime.cloudContext),
+                              isUp
+          ) -> 1,
+          RuntimeHealthMetric(runtime.cloudContext.cloudProvider,
+                              container.imageType,
+                              getRuntimeUI(runtime.labels),
+                              getAzureCloudContext(runtime.cloudContext),
+                              !isUp
+          ) -> 0
+        )
+      }
+      .map(_.combineAll)
+  }
+
+  /** Records and logs a generic AppMetric */
+  private[monitor] def recordMetric[T <: LeoMetric](
+    appMetric: Map[T, Int]
+  )(implicit ev: Ask[F, AppContext]): F[Unit] =
+    appMetric.toList
+      .parTraverseN(parallelism) { case (metric, count) =>
+        for {
+          ctx <- ev.ask
+          _ <- metrics.gauge(
+            metric.name,
+            count,
+            metric.tags
+          )
+          _ <- logger.debug(ctx.loggingCtx)(s"Recorded metric: ${metric.name}, tags: ${metric.tags}, value: ${count}")
+        } yield ()
+      }
+      .void
+
+  private def getRuntimeUI(labels: LabelMap): RuntimeUI =
+    if (labels.contains(Config.uiConfig.terraLabel)) RuntimeUI.Terra
+    else if (labels.contains(Config.uiConfig.allOfUsLabel)) RuntimeUI.AoU
+    else RuntimeUI.Other
+
+  private def getAzureCloudContext(cloudContext: CloudContext): Option[AzureCloudContext] =
+    (config.includeAzureCloudContext, cloudContext) match {
+      case (true, CloudContext.Azure(cc)) => Some(cc)
+      case _                              => None
+    }
+}
+
+case class LeoMetricsMonitorConfig(checkInterval: FiniteDuration, includeAzureCloudContext: Boolean)
+
+sealed trait LeoMetric {
```

Comment:
Little ADT for Leo metrics. A metric has a name; a type (we are always using `gauge` in this class but there are also `counter` and `timer); tags; and a value.